src/tof/pulse.py

```diff
@@ -13,18 +13,198 @@
+            data=sc.array(dims=[t_dim], values=[1.0, 1.0]),
+            coords={
+                t_dim: sc.array(
+                    dims=[t_dim], values=[tmin.value + dt, tmax.value + dt], unit=t_u
+                )
+            },
+        )
+    if p_wav is None:
+        if (wmin is None) and (wmax is None):
+            raise ValueError("Either `p_wav` or `wmin` and `wmax` must be specified.")
+        dw = (wmax - wmin).value / (sampling - 1)
+        p_wav = sc.DataArray(
+            data=sc.array(dims=[w_dim], values=[1.0, 1.0]),
+            coords={
+                w_dim: sc.array(
+                    dims=[w_dim], values=[wmin.value + dw, wmax.value + dw], unit=w_u
+                )
+            },
+        )
+
+    if tmin is None:
+        tmin = p_time.coords[t_dim].min()
+    if tmax is None:
+        tmax = p_time.coords[t_dim].max()
+    if wmin is None:
+        wmin = p_wav.coords[w_dim].min()
+    if wmax is None:
+        wmax = p_wav.coords[w_dim].max()
+
+    time_interpolator = interp1d(p_time, dim=t_dim, fill_value='extrapolate')
+    wav_interpolator = interp1d(p_wav, dim=w_dim, fill_value='extrapolate')
+    x_time = sc.linspace(
+        dim=t_dim,
+        start=tmin.value,
+        stop=tmax.value,
+        num=sampling,
+        unit=tmin.unit,
+    )
+    x_wav = sc.linspace(
+        dim=w_dim,
+        start=wmin.value,
+        stop=wmax.value,
+        num=sampling,
+        unit=wmin.unit,
+    )
+    p_time = time_interpolator(x_time)
+    p_time /= p_time.data.sum()
+    p_wav = wav_interpolator(x_wav)
+    p_wav /= p_wav.data.sum()
+
+    # In the following, random.choice only allows to select from the values listed
+    # in the coordinate of the probability distribution arrays. This leads to data
+    # grouped into spikes and empty in between because the sampling resolution used
+    # in the linear interpolation above is usually kept low for performance.
+    # To make the distribution more uniform, we add some random noise to the chosen
+    # values, which effectively fills in the gaps between the spikes.
+    # Scipy has some methods to sample from a continuous distribution, but they are
+    # prohibitively slow.
+    # See https://docs.scipy.org/doc/scipy/tutorial/stats/sampling.html for more
+    # information.
+    dt = 0.5 * (tmax - tmin).value / (p_time.sizes[t_dim] - 1)
+    dw = 0.5 * (wmax - wmin).value / (p_wav.sizes[w_dim] - 1)
+
+    # Because of the added noise, some values end up being outside the specified range
+    # for the birth times and wavelengths. Using naive clipping leads to pile-up on the
+    # edges of the range. To avoid this, we remove the outliers and resample until we
+    # have the desired number of neutrons.
+    n = 0
+    times = []
+    wavs = []
+    while n < neutrons:
+        size = neutrons - n
+
+        t = np.random.choice(
+            p_time.coords[t_dim].values, size=size, p=p_time.values
+        ) + np.random.uniform(-dt, dt, size=size)
```

Comment:
Have you considered using a normal distribution?

Reply:
You mean for spreading the values? i.e. instead of the `random.uniform`?
I thought of it but decided it would make bumps instead of spreading evenly, and if the sampling is low (e.g. if you end up selecting only a small wavelength range using choppers), you would see them? (a bit like the spikes and the empty bins, but granted to a lesser degree)

Reply:
My thought was concern potentially more artifacts with uniform (and overlapping) boxes, and generally better properties you get with gaussians, since it is basically you you do when smearing, and it approximates many actual physical effects.

Reply:
So the boxes here should not overlap because their ranges should meet exactly in the middle between two values.
But I will give it a go and see what it gives.

Reply:
So you always have constant bin size in the distribution? Is that checked anywhere? Does the `numpy.choice` make that assumption?